File: packages/go/dusk/core/enforcer.go
package core

import (
	"fmt"
	"time"
)

type EnforcerResult struct {
	Enforce bool
	Body    map[string]string
}

func CheckSunset(ep EndpointConfig) EnforcerResult {
	if ep.SunsetAt == nil {
		return EnforcerResult{}
	}
	sunset, err := time.Parse("2006-01-02", *ep.SunsetAt)
	if err != nil || !time.Now().After(sunset) {
		return EnforcerResult{}
	}

	body := map[string]string{
		"error":   "Gone",
		"message": fmt.Sprintf("%s was sunset on %s and is no longer available.", ep.Path, *ep.SunsetAt),
	}
	if ep.Successor != nil {
		body["successor"] = *ep.Successor
	}
	if ep.MigrationDoc != nil {
		body["migration_doc"] = *ep.MigrationDoc
	}
	return EnforcerResult{Enforce: true, Body: body}
}

File: packages/python/src/dusk/core/enforcer.py
from datetime import date
from dusk.core.models import EndpointConfig


class SunsetEnforcer:
    def check(self, endpoint: EndpointConfig) -> tuple[bool, dict | None]:
        """Returns (should_enforce_410, response_body_or_None)."""
        if not endpoint.sunset_at:
            return False, None

        sunset = date.fromisoformat(endpoint.sunset_at)
        if date.today() <= sunset:
            return False, None

        body: dict = {
            "error": "Gone",
            "message": f"{endpoint.path} was sunset on {endpoint.sunset_at} and is no longer available.",
        }
        if endpoint.successor:
            body["successor"] = endpoint.successor
        if endpoint.migration_doc:
            body["migration_doc"] = endpoint.migration_doc

        return True, body
